### PR TITLE
Allow building with cmake (to ease building on windows)

### PR DESCRIPTION
**Pull request by trondn**
The patch series allows for building on windows with the following steps:

```
git clone git://github.com/trondn/snappy
mkdir build
cd build
cmake -G "NMake Makefiles" ..\snappy
nmake all test
```

(It should also build on unix systems by dropping the "-G NMake Makefiles" part)

### CODE REVIEW

**Comment by Optiligence** [comment]
Just my 2 cents when i see this: 
https://cmake.org/cmake/help/v3.4/module/GenerateExportHeader.html
https://cmake.org/cmake/help/v3.4/module/CMakePackageConfigHelpers.html (omitting this disregards an integral part of the cmake ecosystem)
for the vs compile fix see #10

**Comment by trondn** [comment]
I've never used the GenerateExportHeader or PackageConfig thing before :-) Updated the set to use that

**Comment by Nulifier** [comment]
Your config.h.in file doesn't work for MSVC 2015. You need to either check for _MSC_VER <= 1900 or increment it as it still doesn't have ssize_t.

**Comment by jakirkham** [comment]
Would be great if this works on VS 2008 also.

**Comment by ccleve** [comment]
I cloned today's version of google/snappy and applied this pull request to it. It built fine and the unit tests ran fine as well. I'm on Windows 10 with Visual Studio 14. There were a few compile warnings (attached).

If the warnings are not of concern, then please accept this PR. It'll make building projects that depend on snappy easier.
[warnings.txt](https://github.com/google/snappy/files/408113/warnings.txt)

**Comment by rapoth** [comment]
@trondn Thank you for submitting this PR! I can confirm this works on today's version of google/snappy with similar warnings with Visual Studio 14 2015 on Windows 10. 

When I run the unit tests, I get a warning:

> Running microbenchmarks.
> WARNING: Compiled without optimization, will be slow.
> 

Do you by any chance happen to know why?

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,155 @@
+PROJECT(Snappy C CXX)
+CMAKE_MINIMUM_REQUIRED(VERSION 3.0)
+
+SET(SNAPPY_MAJOR 1)
+SET(SNAPPY_MINOR 1)
+SET(SNAPPY_PATCHLEVEL 3)
+
+# I haven't looked at what the soname for the previous
+# versions of snappy... let's just start with 1.1.3
+SET(SNAPPY_SONAME "1.1.3")
```
Comment (vrabaud) [inline CMakeLists.txt]: That variable is useless now that you use SNAPPY_MAJOR in SET_TARGET_PROPERTIES.

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,155 @@
+PROJECT(Snappy C CXX)
```
Comment (vrabaud) [inline CMakeLists.txt]: The project is CXX only no ?

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+INCLUDE(CheckIncludeFiles)
+INCLUDE(CheckLibraryExists)
+INCLUDE(CheckCXXSourceCompiles)
+INCLUDE(TestBigEndian)
+INCLUDE(GenerateExportHeader)
+INCLUDE(CMakePackageConfigHelpers)
+
+TEST_BIG_ENDIAN(WORDS_BIG_ENDIAN)
+IF (WORDS_BIG_ENDIAN)
+    MESSAGE(STATUS "Builing on big endian system")
```
Comment (vrabaud) [inline CMakeLists.txt]: typo: Building

**Comment by vrabaud** [comment]
The three commits should probably be independent PR, not to lose track of the awesomeness of the CMake commit :)

**Comment by vrabaud** [comment]
For the warning in VS, maybe you can add something like:
```
if(NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE "Release" CACHE
    "Build type: Release, Debug or RelWithDebInfo" STRING FORCE
  )
endif()
```

File: CMakeLists.txt (hunk 2 above)
Comment (alkis) [inline CMakeLists.txt]: We export a C only API by wrapping the C++ code. See snappy-c.h.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+CONFIGURE_FILE(${Snappy_SOURCE_DIR}/cmake/config.h.in config.h)
+CONFIGURE_FILE(${Snappy_SOURCE_DIR}/cmake/snappy-stubs-public.h.in snappy-stubs-public.h)
+
+IF (WIN32)
+    ADD_DEFINITIONS(-D_CRT_SECURE_NO_WARNINGS)
+ENDIF (WIN32)
+
+ADD_DEFINITIONS(-DHAVE_CONFIG_H)
+
+INCLUDE_DIRECTORIES(BEFORE ${Snappy_SOURCE_DIR})
```
Comment (alkis) [inline CMakeLists.txt]: Forgive my Cmake noobishness, but Cmake [best practices] (http://www.slideshare.net/DanielPfeifer1/cmake-48475415) says do not use include_directories but target_include_directories instead.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+
+CONFIGURE_FILE(${Snappy_SOURCE_DIR}/cmake/config.h.in config.h)
+CONFIGURE_FILE(${Snappy_SOURCE_DIR}/cmake/snappy-stubs-public.h.in snappy-stubs-public.h)
+
+IF (WIN32)
+    ADD_DEFINITIONS(-D_CRT_SECURE_NO_WARNINGS)
+ENDIF (WIN32)
+
+ADD_DEFINITIONS(-DHAVE_CONFIG_H)
```
Comment (alkis) [inline CMakeLists.txt]: target_add_definitions instead?

File: snappy-c.h [hunk 6]
```diff
@@ -68,6 +69,7 @@ typedef enum {
  *   }
  *   free(output);
  */
+SNAPPY_PUBLIC_API
```
Comment (alkis) [inline snappy-c.h]: Is it possible to remove all SNAPPY_PUBLIC_API and use WINDOWS_EXPORT_ALL_SYMBOLS instead?

**Comment by jakirkham** [comment]
Not to be noisy, but is there any news on this PR?

**Comment by alkis** [comment]
Closing this because CMake support has been added internally and will be part of the next release.

**Pull request closed without merging**